text/0000-ember-data-links-meta-improvements.md

```diff
@@ -0,0 +1,340 @@
+  */
+  value()
+
+  /**
+    Reload the underlying `RecordArray`.
+
+    This will re-fetch the `self` link.
+
+    @return {Promsise} resolving with the reloaded `DS.RecordArray`
+  */
+  reload()
+
+  /**
+    Get all links associated with this `DS.RecordArray`.
+
+    This is useful for pagination for example.
+
+    @return {Array<DS.LinkReference>}
+  */
+  links()
+
+  /**
+    Get the meta associated with the underlying `DS.RecordArray`
+
+    @return {Object}
+  */
+  meta()
+
+}
+```
+
+### `RecordReference.meta()`
+
+The `RecordReference` gets a new `meta()` method, which returns the last
+associated meta for the record pushed into the store via `findRecord()`,
+`queryRecord()`, `store.push()` or `store.pushPayload()`.
+
+```js
+// GET /books/1
+// {
+//   data: { … },
+//   meta: {
+//     isThriller: true
+//   }
+// }
+store.findRecord("book", 1).then(function(book) {
+  // get the RecordReference
+  let bookRef = book.ref();
+
+  let meta = bookRef.meta();
+  assert.equal(meta.isThriller, true);
+});
+```
+
+### `[Record|BelongsTo|HasMany|RecordArray]Reference.links()`
+
+The `links()` method on those references allows to get all link references
+associated with the reference, or the specific `LinkReference`, if the name of
+the link is passed:
+
+```js
+// {
+//   data: {
+//     type: "book",
+//     id: 1,
+//     relationships: {
+//       chapters: {
+//         data: [ … ],
+//         links: {
+//           "next": { … },
+//           "prev": { … }
+//         }
+//       }
+//     }
+//   }
+// }
+let chaptersRef = store.peekRecord("book", 1).hasMany("chapters");
+
+// [<DS.LinkReference>, <DS.LinkReference>]
+let links = chaptersRef.links();
+
+// DS.LinkReference
+let nextLink = chaptersRef.links("next");
+```
+
+### Adaptions to current references / models
+
+To fully complete the circle of references, the existing classes are modified
+as follows:
+
+- `RecordReference`
+  - add `links()`
+- `Model`
+  - add `ref()` to get the corresponding `RecordReference`
+- `BelongsToReference`
+  - add `links()`
+  - add `parentRef()` to get a reference to the parent `RecordReference`
+- `HasManyReference`
+  - add `links()`
+  - add `parentRef()` to get a reference to the parent `RecordReference`
+- `ManyArray`
+  - add `ref()` to get the corresponding `HasManyReference`
+- `AdapterPopulatedRecordArray` and `RecordArray`
+  - add `ref()` to get the corresponding `RecordArrayReference`
+
+### Hooks
+
+Additionally to the new types of references and the extension of the existing
+ones, the Model Lifecycle Hooks proposed in
+[RFC#123](https://github.com/emberjs/rfcs/pull/123) allow to react to changes
+for the underlying data of references. The hooks are explained in greater
+detail in that RFC, but record level meta data handling for example might look
+like this:
+
+```js
+// app/models/books.js
+import Model from "ember-data/model";
+
+const Book = Model.extend({});
+
+Book.reopenClass({
+
+  didReceiveData(recordRef) {
```

Comment:
FWIW, this is very similar to the way we are handling this case in our hacked-up version of Ember Data we are currently using to address this problems.  +1 to the approach from our POV.